src/Features/LanguageServer/CommonLanguageServerProtocol.Framework/LanguageServerTarget.cs

```diff
@@ -0,0 +1,427 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.Contracts;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using StreamJsonRpc;
+
+namespace CommonLanguageServerProtocol.Framework;
+
+public interface ILspServiceProvider
+{
+    ILspServices CreateServices(string serverKind);
+}
+
+public interface ILspServices : IDisposable
+{
+    T GetRequiredService<T>();
+}
+
+public interface IRequestExecutionQueue<RequestContextType> where RequestContextType : struct
```

Comment:
RequestContextType might be better as an interface, and no need to limit to struct.